src/Xmobar/Plugins/StdinReader.hs

```
@@ -22,30 +24,36 @@ import Prelude
 import System.Posix.Process
 import System.Exit
 import System.IO
+import System.IO.Error (isEOFError)
 import Xmobar.Run.Exec
 import Xmobar.X11.Actions (stripActions)
-import Xmobar.System.Utils (onSomeException)
-import Control.Monad (when)
+import Control.Concurrent (threadDelay)
+import Control.Exception
+import Control.Monad (forever)
 
 data StdinReader = StdinReader | UnsafeStdinReader
   deriving (Read, Show)
 
 instance Exec StdinReader where
-  start stdinReader cb = do
-    -- The EOF check is necessary for certain systems
-    -- More details here https://github.com/jaor/xmobar/issues/442
-    eof <- isEOF
-    when eof $
-      do hPrint stderr "xmobar: eof at an early stage"
-         exitImmediately ExitSuccess
-    s <-
-      getLine `onSomeException`
-      (\e -> do
-         let errorMessage = "xmobar: Received exception " <> show e
-         hPrint stderr errorMessage
-         cb errorMessage)
-    cb $ escape stdinReader s
-    start stdinReader cb
+  start stdinReader cb = forever $ (cb . escape stdinReader =<< getLine) `catch` handler
+    where
+      -- rethrow async exceptions like ThreadKilled, etc.
+      handler (fromException -> Just e) = throwIO (e :: SomeAsyncException)
+      -- XMonad.Hooks.DynamicLog.statusBar starts new xmobar on every xmonad
+      -- reload and the old xmobar is only signalled to exit via the pipe
+      -- being closed, so we must unconditionally terminate on EOF, otherwise
+      -- there'd be a pileup of xmobars
+      handler (fromException -> Just e) | isEOFError e = exitImmediately ExitSuccess
+      -- any other exception, like "invalid argument (invalid byte sequence)",
+      -- is logged to both stderr and the bar itself, throttled to avoid
+      -- excessive CPU usage whenever someone pipes garbage into xmobar, and
+      -- then discarded without terminating, so a single charset error doesn't
+      -- break the entire xmobar
+      handler e = do
```

I think if we receive an exception like `"invalid argument (invalid byte sequence)"`, it's a bug somewhere. I remember fixing such a big recently on both xmobar and xmonad-contrib side. So, I'm worried that this handling might bury bugs.

---

Well the error message will still be logged to stderr and will be visible on screen for at least a second, so I wouldn't call that "burying". But I'm not familiar with the specific bug you're talking about. Would there be another line that wouldn't trigger this handler and hide the error message?

---

>  so I wouldn't call that "burying"

Usually I won't go and check `stderr` or `xsession.errors` unless something stops working. But I do think this improves the current situation! :-)

>  But I'm not familiar with the specific bug you're talking about.

* https://github.com/xmonad/xmonad-contrib/pull/334
* https://github.com/xmonad/xmonad-contrib/pull/346

---

> Usually I won't go and check stderr or xsession.errors unless something stops working. But I do think this improves the current situation! :-)

Well, what we could do is to show the error but also keep reading from stdin (to be able to detect EOF an terminate) and just discard whatever we read. Or a mix of both: always show the last error, if any, and append the line from stdin (if decodable, obviously) to it. I'm not sure if all this is worth it but it can certainly be done. :-)

---

i'm not sure either.  is that something that we could add later?

---

Anything can be added later. :-)